python_modules/dagster-pipes/dagster_pipes/__init__.py

```diff
@@ -18,27 +19,52 @@
     Any,
     ClassVar,
     Dict,
+    FrozenSet,
     Generic,
     Iterable,
     Iterator,
-    Literal,
     Mapping,
     Optional,
     Sequence,
     Set,
     TextIO,
     Type,
-    TypedDict,
     TypeVar,
     Union,
     cast,
-    final,
-    get_args,
 )
 
+if sys.version_info >= (3, 8):
+    from typing import Literal, TypedDict, final
+else:
+    try:
+        from typing_extensions import Literal, TypedDict, final
+    except ImportError:
+        raise Exception(
+            "Error importing from `typing_extensions`. The minimum"
+            " official version for dagster-pipes is 3.8, but on lower"
+            " Python versions dagster-pipes will attempt to import "
+            " certain typing constructs using the `typing-extensions` "
+            " package. If you are using a Python version lower than 3.8,"
+            " please make sure `typing-extensions` is installed in"
+            " your environment."
+        )
+
+
 if TYPE_CHECKING:
     from unittest.mock import MagicMock
 
+
+def _literal_to_frozenset(literal: Any) -> FrozenSet[Any]:
+    if sys.version_info >= (3, 8):
+        from typing import get_args
+
+        return frozenset(get_args(literal))
+    else:
+        elements = re.search(r"Literal\[(.*)\]", repr(literal))[1].split(", ")
+        return frozenset([e.strip("\"'") for e in elements])
```

Comment:
This seems crazy. Can we not construct the `Literal` from a set instead?

```python

_METADATA_TYPES = {"__infer__", "text", "url"...}
PipesMetadataType = Literal(list(_METADATA_TYPES))
```

Alternatively you can have the types collection and the Literal declared in parallel with the list duplicated and write a unit test to keep them up-to-date. This regex thing is a non-starter.

Reply:
>Can we not construct the Literal from a set instead?

You cannot do this because the purpose of `Literal` is static analysis, so it can't use evaluated expressions.

However I was able to use a different method of direct access to `Literal.__values__`, which apparently works on 3.6.